src/main/kotlin/org/dif/peerdid/PeerDIDUtils.kt

```diff
@@ -0,0 +1,149 @@
+import java.nio.ByteBuffer
+import java.util.HashMap
+
+internal fun buildDIDDocNumalgo0(peerDID: PeerDID): String {
+    val inceptionKey = peerDID.substring(11)
+    val encodingAlgorithm = peerDID[10]
+
+    if (!isInEncodingTypes(encodingAlgorithm))
+        throw IllegalArgumentException("Unsupported encoding algorithm of key: $encodingAlgorithm")
+
+    val decodedEncnumbasis = decodeEncnumbasis(inceptionKey, peerDID)
+    val gson = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
+    val diddoc = mapOf(
+        "id" to peerDID,
+        "authentication" to gson.toJsonTree(decodedEncnumbasis)
+    )
+    return gson.toJson(diddoc)
+}
+
+internal fun buildDIDDocNumalgo2(peerDID: PeerDID): String {
+    val keys = peerDID.drop(11)
+    val keysList = keys.split(".")
+    var service = ""
+    val keysWithoutPurposeCode = mutableListOf<String>()
+    for (key in keysList) {
+        if (key[0] != 'S') {
+            val transform = key[1]
+            if (!isInEncodingTypes(transform))
+                throw IllegalArgumentException("Unsupported transform part of PeerDID: $transform")
+
+            keysWithoutPurposeCode.add(key.drop(2))
+        } else service = key.drop(1)
+    }
+    val decodedEncnumbasises = keysWithoutPurposeCode.map { key -> decodeEncnumbasis(key, peerDID) }
+    val decodedService = decodeService(service, peerDID)
+
+    val authentication = mutableListOf<Map<String, String>>()
+    val keyAgreement = mutableListOf<Map<String, String>>()
+
+    for (i in decodedEncnumbasises.indices) {
+        if (PublicKeyTypeAuthentication.values().any { value -> value.name == decodedEncnumbasises[i]["type"] })
+            authentication.add(decodedEncnumbasises[i])
+        else if (PublicKeyTypeAgreement.values().any { value -> value.name == decodedEncnumbasises[i]["type"] })
+            keyAgreement.add(decodedEncnumbasises[i])
+        else
+            throw IllegalArgumentException("Invalid key type of: ${keysList[i]}")
+    }
+
+    val gson = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
+    val diddoc = mapOf(
+        "id" to peerDID,
+        "authentication" to gson.toJsonTree(authentication),
+        "keyAgreement" to gson.toJsonTree(keyAgreement),
+        "service" to gson.toJsonTree(decodedService)
+    )
+    return gson.toJson(diddoc)
+}
+
+private fun isInEncodingTypes(encodingAlgorithm: Char): Boolean {
+    return EncodingType.values().any { type -> type.type == encodingAlgorithm }
+}
+
+private fun decodeService(encodedService: JSON, peerDID: PeerDID): Map<String, String> {
+    val decodedService = Base64.decodeBase64(encodedService).decodeToString()
+    val gson = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
+    val serviceMap = gson.fromJson(decodedService, HashMap::class.java)
+    val serviceType = serviceMap.remove("t").toString().replace("dm", "didcommmessaging")
+
+    return mapOf(
+        "id" to peerDID.plus('#').plus(serviceType),
```

Comment:
What if we have multiple services (array)? They all will have the same `id`? If so, that must be fixed, `id` must be unique. 
Probably we need to check it for peer-did python as well.